### PR TITLE
feat: Add the `floyd_cycle_detection_algo.cpp` file/algorithm

### PR DESCRIPTION
**Pull request by Swastyy**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1535"></a>

### CODE REVIEW

**Comment by Swastyy** [comment]
@Panquesito7 Check if this will work for #1529

**Review by mishraabhinn — COMMENTED** [review]
Why there are multiple files again? One file one pr.

**Comment by Swastyy** [comment]
> Why there are multiple files again? One file one pr.

They aren't mine (iterative_tree_traversal.cpp, graph_coloring.cpp and random_pivot_quick_sort.cpp).

**Comment by mishraabhinn** [comment]
> > Why there are multiple files again? One file one pr.
> 
> They aren't mine (iterative_tree_traversal.cpp, graph_coloring.cpp and random_pivot_quick_sort.cpp).
How this is even possible. 
Screenshot 2021-07-19 at 17 38 29

**Comment by Swastyy** [comment]
Okay There are some conflicts between master branch of my forked repo and the real one. I will fork it again but @mishraabhinn merge this last PR #1527 from old fork

**Comment by mishraabhinn** [comment]
> Okay There are some conflicts between master branch of my forked repo and the real one. I will fork it again but @mishraabhinn merge this last PR #1527 from old fork

There are also 5 files. I don't understand. 
[Branch can stay updated with master ](https://gist.github.com/whoisryosuke/36b3b41e738394170b9a7c230665e6b9)
[For making forked repo updated](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/configuring-a-remote-for-a-fork)

**Comment by mishraabhinn** [comment]
> > Okay There are some conflicts between master branch of my forked repo and the real one. I will fork it again but @mishraabhinn merge this last PR #1527 from old fork
> 
> There are also 5 files. I don't understand.
> [Branch can stay updated with master ](https://gist.github.com/whoisryosuke/36b3b41e738394170b9a7c230665e6b9)
> [For making forked repo updated](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/configuring-a-remote-for-a-fork)

I think this is happening because you create branches from previous branch. I advice you that when you create a new branch go to master update master then create a new branch.

**Comment by Swastyy** [comment]
> > > Okay There are some conflicts between master branch of my forked repo and the real one. I will fork it again but @mishraabhinn merge this last PR #1527 from old fork
> > 
> > 
> > There are also 5 files. I don't understand.
> > [Branch can stay updated with master ](https://gist.github.com/whoisryosuke/36b3b41e738394170b9a7c230665e6b9)
> > [For making forked repo updated](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/configuring-a-remote-for-a-fork)
> 
> I think this is happening because you create branches from previous branch. I advice you that when you create a new branch go to master update master then create a new branch.

Ha I think so

**Comment by Swastyy** [comment]
> > Okay There are some conflicts between master branch of my forked repo and the real one. I will fork it again but @mishraabhinn merge this last PR #1527 from old fork
> 
> There are also 5 files. I don't understand.
> [Branch can stay updated with master ](https://gist.github.com/whoisryosuke/36b3b41e738394170b9a7c230665e6b9)
> [For making forked repo updated](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/configuring-a-remote-for-a-fork)

Master branch is anyways the default branch of forked repo but if I can update a particular branch from original repo that will be great. Thanks for the links. 👍🏻

**Comment by Panquesito7** [comment]
All other files are changed automatically by the workflow, so even if you undo those changes, they'll come back. @mishraabhinn.

File: search/floyd_cycle_detection_algo.cpp [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+ * integer is in the range [1, n] inclusive. If there is only one duplicate
+ * number in the input array, this algorithm returns the duplicate number in
+ * O(1) space and the time complexity is less than O(n^2) without modifying the
+ * original array, otherwise, it returns -1.
+ * @author [Swastika Gupta](https://github.com/Swastyy)
+ */
+
+#include <cassert>   /// for assert
+#include <iostream>  /// for IO operations
+#include <vector>    /// for std::vector
```
Comment (Panquesito7) [inline search/floyd_cycle_detection_algo.cpp:15]:
```suggestion
#include <vector>    /// for std::vector

```

File: search/floyd_cycle_detection_algo.cpp [hunk 2]
```diff
@@ -0,0 +1,96 @@
+/**
+ * @file
+ * @brief Implementation of [Floyd's Cycle
+ * Detection](https://en.wikipedia.org/wiki/Cycle_detection) algorithm
+ * @details Given an array of integers containing 'n + 1' integers, where each
```
Comment (Panquesito7) [inline search/floyd_cycle_detection_algo.cpp:5]:
```suggestion
 * @details
 * Given an array of integers containing 'n + 1' integers, where each
```

File: search/floyd_cycle_detection_algo.cpp [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+}
+}  // namespace cycle_detection
+}  // namespace search
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+    // Test 1
```
Comment (Panquesito7) [inline search/floyd_cycle_detection_algo.cpp:63]:
```suggestion
    // 1st test
```

File: search/floyd_cycle_detection_algo.cpp [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    // [3, 4, 8, 5, 9, 1, 2, 6, 7, 4] return 4
+    std::vector<uint32_t> array1 = {3, 4, 8, 5, 9, 1, 2, 6, 7, 4};
+    std::cout << "Test 1... ";
+    assert(search::cycle_detection::duplicateNumber(array1, array1.size()) ==
+           4);  // here the duplicate number is 4
+    std::cout << "passed" << std::endl;
+
+    // Test 2
```
Comment (Panquesito7) [inline search/floyd_cycle_detection_algo.cpp:71]:
```suggestion
    // 2nd test
```

File: search/floyd_cycle_detection_algo.cpp [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    // [1, 2, 3, 4, 2] return 2
+    std::vector<uint32_t> array2 = {1, 2, 3, 4, 2};
+    std::cout << "Test 2... ";
+    assert(search::cycle_detection::duplicateNumber(array2, array2.size()) ==
+           2);  // here the duplicate number is 2
+    std::cout << "passed" << std::endl;
+
+    // Test 3
```
Comment (Panquesito7) [inline search/floyd_cycle_detection_algo.cpp:79]:
```suggestion
    // 3rd test
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work! 😄:+1:

**Comment by mishraabhinn** [comment]
> All other files are changed automatically by the workflow, so even if you undo those changes, they'll come back. @mishraabhinn.
> 
> 

I don’t get this. How automatically?

**Comment by Panquesito7** [comment]
> > All other files are changed automatically by the workflow, so even if you undo those changes, they'll come back. @mishraabhinn.
> 
> I don’t get this. How automatically?

https://github.com/TheAlgorithms/C-Plus-Plus/pull/1535/commits/b8dd0a6aaeb56ce7f9c8372146bde518c9ef7376
https://github.com/TheAlgorithms/C-Plus-Plus/pull/1535/commits/927ff99ef062f391c60036dee516caca5004800f
https://github.com/TheAlgorithms/C-Plus-Plus/pull/1535/commits/ddaba3b77092202cc831726870d33add736872ba
https://github.com/TheAlgorithms/C-Plus-Plus/pull/1535/commits/f011de038b86530c28987f4d443d57a45ccdba7d
https://github.com/TheAlgorithms/C-Plus-Plus/pull/1535/commits/2c4cb52b4288bb34cf3abd8bb0c8909d4ae1e5f2

File: search/floyd_cycle_detection_algo.cpp [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+    // 2nd test
+    // [1, 2, 3, 4, 2] return 2
+    std::vector<uint32_t> array2 = {1, 2, 3, 4, 2};
+    std::cout << "Test 2... ";
+    assert(search::cycle_detection::duplicateNumber(array2, array2.size()) ==
+           2);  // here the duplicate number is 2
+    std::cout << "passed" << std::endl;
+
+    // Test 3
```
Comment (Panquesito7) [inline search/floyd_cycle_detection_algo.cpp:81]:
```suggestion
    // 3rd test
```

**Review by Panquesito7 — COMMENTED** [review]
Almost there! 😄

**Review by Panquesito7 — APPROVED** [review]
LGTM. Thank you for your contribution! 😄🎉

**Pull request closed without merging**